src/modules/atomic-queue/AtomicQueue.sol

```diff
@@ -0,0 +1,298 @@
+
+    /**
+     * @notice Maps user address to offer asset to want asset to a AtomicRequest struct.
+     */
+    mapping(address => mapping(ERC20 => mapping(ERC20 => AtomicRequest))) public userAtomicRequest;
+
+    //============================== ERRORS ===============================
+
+    error AtomicQueue__UserRepeated(address user);
+    error AtomicQueue__RequestDeadlineExceeded(address user);
+    error AtomicQueue__UserNotInSolve(address user);
+    error AtomicQueue__ZeroOfferAmount(address user);
+
+    //============================== EVENTS ===============================
+
+    /**
+     * @notice Emitted when `updateAtomicRequest` is called.
+     */
+    event AtomicRequestUpdated(
+        address user,
+        address offer,
+        address want,
+        uint256 amount,
+        uint256 deadline,
+        uint256 minPrice,
+        uint256 timestamp
+    );
+
+    /**
+     * @notice Emitted when `solve` exchanges a users shares for the underlying asset.
+     */
+    event AtomicRequestFulfilled(
+        address user,
+        address offer,
+        address want,
+        uint256 sharesSpent,
+        uint256 assetsReceived,
+        uint256 timestamp
+    );
+
+    //============================== USER FUNCTIONS ===============================
+
+    /**
+     * @notice Get a users Withdraw Request.
+     */
+    function getUserAtomicRequest(address user, ERC20 offer, ERC20 want) external view returns (AtomicRequest memory) {
+        return userAtomicRequest[user][offer][want];
+    }
+
+    /**
+     * @notice Helper function that returns either
+     *         true: Withdraw request is valid.
+     *         false: Withdraw request is not valid.
+     * @dev It is possible for a withdraw request to return false from this function, but using the
+     *      request in `updateAtomicRequest` will succeed, but solvers will not be able to include
+     *      the user in `solve` unless some other state is changed.
+     */
+    function isAtomicRequestValid(
+        ERC20 offer,
+        address user,
+        AtomicRequest calldata userRequest
+    ) external view returns (bool) {
+        // Validate amount.
+        if (userRequest.offerAmount > offer.balanceOf(user)) return false;
+        // Validate deadline.
+        if (block.timestamp > userRequest.deadline) return false;
+        // Validate approval.
+        if (offer.allowance(user, address(this)) < userRequest.offerAmount) return false;
+        // Validate offerAmount is nonzero.
+        if (userRequest.offerAmount == 0) return false;
+        // Validate atomicPrice is nonzero.
+        if (userRequest.atomicPrice == 0) return false;
+
+        return true;
+    }
+
+    /**
+     * @notice Allows user to add/update their withdraw request.
+     * @notice It is possible for a withdraw request with a zero atomicPrice to be made, and solved.
+     *         If this happens, users will be selling their shares for no assets in return.
```

Comment:
Because it is very difficult to keep this permissionless, and gas efficient if we try to enforce some minimum atomicPrice. Especially in this setup where the offer/want could be any asset. 
If we tried to validate the atomic price, then we would need to be able to price both assets using the price router, which makes it more difficult to support exotic assets like yETH. Also we would need an owner that would then set the minimum price multiplier. Which even that is tricky because maybe a user wants to have a request filled immediately so they opt to set an atomic price that is 10% lower than market price. If we enforce that the atomic price can only be 5% lower than market price, then we are taking more freedom from the user.

Reply:
Thank you for the explanation, this makes more sense now to me. I can see justification for both sides, but in efforts to keep this as a permissionless tool that any actor can engage with, it makes sense to leave it as is. This becomes a permissionless trade agreement service provision to users, solvers, revolving around cellars / assets, essentially.